File: algorithm/sort.cpp
#include <iostream>
#include <vector>
#include <random>
#include <chrono>
#include <limits>
#include <numeric>
#include <functional>
#include <string>
#include <iomanip>
#include <stdexcept>
#include <cmath>

#define VARIABLE_TO_STRING(x) (#x)

typedef int64_t index_type;

template <typename T>
std::ostream &operator<<(std::ostream &os, const std::vector<T> &v)
{
    os << "{ ";
    for (auto it = v.begin(); it != v.end(); ++it)
    {
        if (it != v.begin())
        {
            os << ", ";
        }
        os << *it;
    }
    os << " }";

    return os;
}

std::vector<int> generate(size_t num)
{
    if (num == 0)
    {
        return {};
    }

    std::random_device rd;
    std::mt19937 engine(rd());
    std::uniform_int_distribution<int> uniform(0, std::numeric_limits<int16_t>::max());
    std::vector<int> array(num, 0);

    for (size_t i = 0; i < num; ++i)
    {
        array[i] = uniform(engine);
    }

    return array;
}

void bubble_sort(std::vector<int> &array)
{
    index_type length = static_cast<index_type>(array.size());
    if (length < 2)
    {
        return;
    }

    for (index_type i = 0; i < length - 1; ++i)
    {
        for (index_type j = 0; j < length - 1 - i; ++j)
        {
            if (array[j] > array[j + 1])
            {
                std::swap(array[j], array[j + 1]);
            }
        }
    }
}

void select_sort(std::vector<int> &array)
{
    index_type length = static_cast<index_type>(array.size());
    if (length < 2)
    {
        return;
    }

    for (index_type i = 0; i < length; i++)
    {
        index_type index = i;

        for (index_type j = i + 1; j < length; ++j)
        {
            if (array[j] < array[index])
            {
                index = j;
            }
        }

        if (index != i)
        {
            std::swap(array[index], array[i]);
        }
    }
}

void insert_sort(std::vector<int> &array)
{
    index_type length = static_cast<index_type>(array.size());
    if (length < 2)
    {
        return;
    }

    for (index_type i = 0; i < length; ++i)
    {
        index_type index = i;
        auto tmp = array[i];

        while (index > 0 && array[index - 1] > tmp)
        {
            array[index] = array[index - 1];
            --index;
        }

        array[index] = tmp;
    }
}

void shell_sort(std::vector<int> &array)
{
    index_type length = static_cast<index_type>(array.size());
    if (length < 2)
    {
        return;
    }

    for (index_type span = length / 2; span > 0; span /= 2)
    {
        for (index_type i = span; i < length; ++i)
        {
            index_type index = i;
            auto tmp = array[i];

            while (index >= span && array[index - span] > tmp)
            {
                array[index] = array[index - span];
                index -= span;
            }

            array[index] = tmp;
        }
    }
}

void heap_sort(std::vector<int> &array)
{
    index_type length = static_cast<index_type>(array.size());
    if (length < 2)
    {
        return;
    }

    std::function<void(std::vector<int> &, index_type, index_type)> build_heap =
        [&build_heap](std::vector<int> &array, index_type length, index_type index)
    {
        index_type father = index;
        index_type lchild = 2 * index + 1;
        index_type rchild = 2 * index + 2;

        // 找到比子节点中比父节点大的值，并与之交换，然后向下递推
        if (lchild < length && array[father] < array[lchild])
        {
            father = lchild;
        }
        if (rchild < length && array[father] < array[rchild])
        {
            father = rchild;
        }

        if (father != index)
        {
            std::swap(array[father], array[index]);
            build_heap(array, length, father);
        }
    };

    for (int i = length - 1; i >= 0; --i)
    {
        build_heap(array, length, i);
    }

    for (int i = length - 1; i >= 0; --i)
    {
        std::swap(array[i], array[0]);
        build_heap(array, i, 0);
    }
}

void merge_sort(std::vector<int> &array)
{
    index_type length = static_cast<index_type>(array.size());
    if (length < 2)
    {
        return;
    }

    std::vector<int> tmp(length);

    std::function<void(std::vector<int> &, index_type, index_type, index_type)> merge =
        [&tmp](std::vector<int> &array, index_type left, index_type mid, index_type right)
    {
        index_type lIdx = left;
        index_type rIdx = mid + 1;
        index_type index = 0;

        // 将数组分为两组，按大小归并入原数组
        while (lIdx <= mid && rIdx <= right)
        {
            if (array[lIdx] <= array[rIdx])
            {
                tmp[index++] = array[lIdx++];
            }
            else
            {
                tmp[index++] = array[rIdx++];
            }
        }

        while (lIdx <= mid)
        {
            tmp[index++] = array[lIdx++];
        }

        while (rIdx <= right)
        {
            tmp[index++] = array[rIdx++];
        }

        index = 0;
        while (left <= right)
        {
            array[left++] = tmp[index++];
        }
    };
    std::function<void(std::vector<int> &, index_type, index_type)> mg_sort =
        [&merge, &mg_sort](std::vector<int> &array, index_type left, index_type right)
    {
        if (left < right)
        {
            index_type mid = left + ((right - left) >> 1);
            mg_sort(array, left, mid);
            mg_sort(array, mid + 1, right);
            merge(array, left, mid, right);
        }
    };

    mg_sort(array, 0, length - 1);
}

void quick_sort(std::vector<int> &array)
{
    index_type length = static_cast<index_type>(array.size());
    if (length < 2)
    {
        return;
    }

    std::default_random_engine engine(std::chrono::system_clock::now().time_since_epoch().count());
    std::uniform_int_distribution<index_type> uniform;
    std::function<index_type(std::vector<int> &, index_type, index_type)> partition =
        [&engine, &uniform](std::vector<int> &array, index_type left, index_type right) -> index_type
    {
        index_type stdIdx = uniform(engine, decltype(uniform)::param_type{left, right});
        std::swap(array[right], array[stdIdx]);

        int standard = array[right];
        index_type index = left;

        for (index_type i = left; i < right; ++i)
        {
            if (array[i] < standard)
            {
                std::swap(array[i], array[index++]);
            }
        }
        std::swap(array[index], array[right]);

        return index;
    };
    std::function<void(std::vector<int> &, index_type, index_type)> qc_sort =
        [&partition, &qc_sort](std::vector<int> &array, index_type left, index_type right)
    {
        if (left < right)
        {
            index_type index = partition(array, left, right);
            qc_sort(array, left, index - 1);
            qc_sort(array, index + 1, right);
        }
    };

    qc_sort(array, 0, length - 1);
}

void count_sort(std::vector<int> &array)
{
    index_type length = static_cast<index_type>(array.size());
    if (length < 2)
    {
        return;
    }

    int maxValue = *std::max_element(array.begin(), array.end());
    std::vector<int> count(maxValue + 1, 0);

    // 记录每一个数字出现的次数
    for (auto i : array)
    {
        ++count[i];
    }

    index_type countSize = static_cast<index_type>(count.size());

    // 写入到原始的数组
    for (index_type i = 0, index = 0; i < countSize; ++i)
    {
        for (index_type j = 0; j < count[i]; ++j)
        {
            array[index++] = i;
        }
    }
}

void radix_sort(std::vector<int> &array)
{
    index_type length = static_cast<index_type>(array.size());
    if (length < 2)
    {
        return;
    }

    int maxValue = *std::max_element(array.begin(), array.end());
    index_type digit = std::floor(log10(maxValue)) + 1;
    std::vector<std::vector<int>> radix(10);

    // 从个位数开始，将每个数放置于不同的基数上
    for (index_type d = 0; d < digit; ++d)
    {
        for (auto i : array)
        {
            double ratio = std::pow(10, d);
            index_type index = i % static_cast<index_type>(ratio * 10) / ratio;
            radix[index].emplace_back(i);
        }

        index_type index = 0;
        for (auto &rad : radix)
        {
            for (auto i : rad)
            {
                array[index++] = i;
            }
            rad.clear();
        }
    }
}

void bucket_sort(std::vector<int> &array)
{
    index_type length = static_cast<index_type>(array.size());
    if (length < 2)
    {
        return;
    }

    auto minmax = std::minmax_element(array.begin(), array.end());
    int minValue = *minmax.first;
    int maxValue = *minmax.second;
    index_type bkSize = (maxValue - minValue) / length + 1;
    std::vector<std::vector<int>> bucket(bkSize);

    // 将每个元素放置到指定的桶中
    for (auto i : array)
    {
        index_type index = (i - minValue) / length;
        bucket[index].emplace_back(i);
    }

    // 对每个桶分别进行排序
    for (auto &bk : bucket)
    {
        std::sort(bk.begin(), bk.end());
    }

    index_type index = 0;

    // 写入到原始的数组
    for (auto &bk : bucket)
    {
        for (auto i : bk)
        {
            array[index++] = i;
        }
    }
}

void std_sort(std::vector<int> &array)
{
    std::sort(array.begin(), array.end());
}

int main()
{
    try
    {
        std::vector<std::pair<std::string, std::function<void(std::vector<int> &)>>> vFunc = {
            {VARIABLE_TO_STRING(bubble_sort), bubble_sort},
            {VARIABLE_TO_STRING(select_sort), select_sort},
            {VARIABLE_TO_STRING(insert_sort), insert_sort},
            {VARIABLE_TO_STRING(shell_sort), shell_sort},
            {VARIABLE_TO_STRING(heap_sort), heap_sort},
            {VARIABLE_TO_STRING(merge_sort), merge_sort},
            {VARIABLE_TO_STRING(quick_sort), quick_sort},
            {VARIABLE_TO_STRING(count_sort), count_sort},
            {VARIABLE_TO_STRING(radix_sort), radix_sort},
            {VARIABLE_TO_STRING(bucket_sort), bucket_sort},
            {VARIABLE_TO_STRING(std_sort), std_sort},
        };

        const auto original_array = generate(10000);
        for (auto &func : vFunc)
        {
            auto array = original_array;
            int64_t start = std::chrono::high_resolution_clock::now().time_since_epoch().count();
            func.second(array);
            int64_t end = std::chrono::high_resolution_clock::now().time_since_epoch().count();

            std::cout << std::setw(12) << std::right << func.first << std::setw(15) << std::right << std::fixed << std::setprecision(3)
                      << static_cast<double>(end - start) / 1000000 << std::setw(8) << std::right
                      << (std::is_sorted(array.begin(), array.end()) ? "true" : "false") << std::endl;
        }
    }
    catch (const std::exception &e)
    {
        std::cerr << e.what() << std::endl;
    }

    return 0;
}

File: algorithm/sort.ts
// 每次从第一个开始，⽐较相邻的元素。如果第⼀个⽐第⼆个⼤，就交换他们两个
function bubbleSort(arr: number[]) {
    for (let i = 0; i < arr.length - 1; i++) {
        for (let j = 0; j < arr.length - 1 - i; j++) {
            if (arr[j] > arr[j + 1]) {
                [arr[j], arr[j + 1]] = [arr[j + 1], arr[j]];
            }
        }
    }
}

// 每一次从待排序的数据元素中选出最小的⼀个元素，存放在序列的起始位置
function selectSort(arr: number[]) {
    for (let i = 0; i < arr.length - 1; i++) {
        let minIndex = i;
        for (let j = i + 1; j < arr.length; j++) {
            if (arr[j] < arr[minIndex]) {
                minIndex = j;
            }
        }
        [arr[i], arr[minIndex]] = [arr[minIndex], arr[i]];
    }
}

// 将⼀个数据插入到已经排好序的有序数据中，从而得到⼀个新的、个数加⼀的有序数据
function insertSort(arr: number[]) {
    for (let i = 1; i < arr.length; i++) {
        let current = arr[i];
        let j = i - 1;
        while (j >= 0 && arr[j] > current) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = current;
    }
}

// 希尔排序是插入排序的改进版，它通过设置一个增量来分组排序，最后再进行一次插入排序
function shellSort(arr: number[]) {
    let gap = Math.floor(arr.length / 2);
    while (gap > 0) {
        for (let i = gap; i < arr.length; i++) {
            let current = arr[i];
            let j = i - gap;
            while (j >= 0 && arr[j] > current) {
                arr[j + gap] = arr[j];
                j -= gap;
            }
            arr[j + gap] = current;
        }
        gap = Math.floor(gap / 2);
    }
}

// 堆排序是将数组看作一棵完全二叉树，将树调整为堆，然后依次将堆顶元素与末尾元素交换，再调整堆
function heapSort(arr: number[]) {
    const length = arr.length;

    const buildHeap = (arr: number[], length: number, index: number): void => {
        let father = index;
        const left = 2 * index + 1;
        const right = 2 * index + 2;

        if (left < length && arr[left] > arr[father]) {
            father = left;
        }

        if (right < length && arr[right] > arr[father]) {
            father = right;
        }

        if (father !== index) {
            [arr[index], arr[father]] = [arr[father], arr[index]];
            buildHeap(arr, length, father);
        }
    };

    for (let i = Math.floor(length / 2) - 1; i >= 0; i--) {
        buildHeap(arr, length, i);
    }

    for (let i = length - 1; i >= 0; i--) {
        [arr[0], arr[i]] = [arr[i], arr[0]];
        buildHeap(arr, i, 0);
    }
}

// 归并排序是将数组分成两部分，然后对两部分进行排序，最后将两部分合并
function mergeSort(arr: number[]) {
    const length = arr.length;
    const tmpArr = new Array(length);

    const merge = (arr: number[], left: number, mid: number, right: number): void => {
        let i = left, j = mid + 1, k = 0;

        while (i <= mid && j <= right) {
            if (arr[i] <= arr[j]) {
                tmpArr[k] = arr[i];
                i++;
            } else {
                tmpArr[k] = arr[j];
                j++;
            }
            k++;
        }

        while (i <= mid) {
            tmpArr[k] = arr[i];
            i++;
            k++;
        }

        while (j <= right) {
            tmpArr[k] = arr[j];
            j++;
            k++;
        }

        for (k = 0; left <= right;) {
            arr[left] = tmpArr[k];
            left++;
            k++;
        }
    };

    const mgSort = (arr: number[], left: number, right: number): void => {
        if (left < right) {
            const mid = left + ((right - left) >> 1);
            mgSort(arr, left, mid);
            mgSort(arr, mid + 1, right);
            merge(arr, left, mid, right);
        }
    };

    mgSort(arr, 0, length - 1);
}

// 快速排序是选择一个基准元素，将数组分成两部分，一部分比基准元素小，一部分比基准元素大，然后递归地对两部分进行排序
function quickSort(arr: number[]) {
    const length = arr.length;

    const partition = (arr: number[], left: number, right: number): number => {
        const random = Math.floor(Math.random() * (right - left + 1)) + left;
        [arr[right], arr[random]] = [arr[random], arr[right]];

        const standard = arr[right];
        let index = left;

        for (let i = left; i < right; i++) {
            if (arr[i] < standard) {
                [arr[i], arr[index]] = [arr[index], arr[i]];
                index++;
            }
        }
        [arr[index], arr[right]] = [arr[right], arr[index]];

        return index;
    };

    const qcSort = (arr: number[], left: number, right: number): void => {
        if (left < right) {
            const index = partition(arr, left, right);
            qcSort(arr, left, index - 1);
            qcSort(arr, index + 1, right);
        }
    };

    qcSort(arr, 0, length - 1);
}

// 计数排序是将数组中的元素作为键存储在额外开辟的数组空间中，然后依次将键值对中的元素放回原数组
function countSort(arr: number[]) {
    const length = arr.length;
    if (length < 2) {
        return;
    }

    let max = arr[0];
    for (let i = 1; i < length; i++) {
        if (arr[i] > max) {
            max = arr[i];
        }
    }

    const tmpArr = new Array(max + 1).fill(0);
    for (const num of arr) {
        tmpArr[num]++;
    }

    let index = 0;
    for (let i = 0; i < tmpArr.length; i++) {
        const count = tmpArr[i];
        for (let j = 0; j < count; j++) {
            arr[index] = i;
            index++;
        }
    }
}

// 基数排序是根据元素的每一位进行排序，从低位到高位依次排序
function radixSort(arr: number[]) {
    const length = arr.length;
    if (length < 2) {
        return;
    }

    let max = arr[0];
    for (let i = 1; i < length; i++) {
        if (arr[i] > max) {
            max = arr[i];
        }
    }

    const digit = Math.floor(Math.log10(max)) + 1;
    const radix: number[][] = Array.from({ length: 10 }, () => []);

    for (let d = 0; d < digit; d++) {
        for (const num of arr) {
            const ratio = Math.pow(10, d);
            const index = Math.floor(num / ratio) % 10;
            radix[index].push(num);
        }

        let index = 0;
        for (let i = 0; i < radix.length; i++) {
            for (const num of radix[i]) {
                arr[index] = num;
                index++;
            }
            radix[i] = [];
        }
    }
}

// 桶排序是根据元素的值将元素分配到不同的桶中，然后对每个桶进行排序，最后将桶中的元素依次放回原数组
function bucketSort(arr: number[]) {
    const length = arr.length;
    if (length < 2) {
        return;
    }

    let max = arr[0];
    let min = arr[0];
    for (let i = 1; i < length; i++) {
        if (arr[i] > max) {
            max = arr[i];
        } else if (arr[i] < min) {
            min = arr[i];
        }
    }

    const bucketSize = Math.floor((max - min) / length) + 1;
    const bucket: number[][] = Array.from({ length: bucketSize }, () => []);

    for (const num of arr) {
        const index = Math.floor((num - min) / length);
        bucket[index].push(num);
    }

    for (let i = 0; i < bucket.length; i++) {
        bucket[i].sort((a, b) => a - b);
    }

    let index = 0;
    for (const b of bucket) {
        for (const num of b) {
            arr[index] = num;
            index++;
        }
    }
}

// 标准排序
function stdSort(arr: number[]) {
    arr.sort((a, b) => a - b);
}

function generate(num: number): number[] {
    return Array.from({ length: num }, () => Math.floor(Math.random() * 1000000));
}

function isSorted(array: number[]): boolean {
    for (let i = 1; i < array.length; i++) {
        if (array[i - 1] > array[i]) {
            return false;
        }
    }
    return true;
}

function main() {
    const vFunc: Array<{ name: string, func: (array: number[]) => void }> = [
        { name: 'bubbleSort', func: bubbleSort },
        { name: 'selectSort', func: selectSort },
        { name: 'insertSort', func: insertSort },
        { name: 'shellSort', func: shellSort },
        { name: 'heapSort', func: heapSort },
        { name: 'mergeSort', func: mergeSort },
        { name: 'quickSort', func: quickSort },
        { name: 'countSort', func: countSort },
        { name: 'radixSort', func: radixSort },
        { name: 'bucketSort', func: bucketSort },
        { name: 'stdSort', func: stdSort },
    ];

    const originalArray = generate(10000);

    vFunc.forEach(({ name, func }) => {
        const array = [...originalArray];
        const start = performance.now();
        func(array);
        const end = performance.now();

        console.log(
            `${name.padStart(12)} ${(end - start).toFixed(3).padStart(15)} ${(isSorted(array) ? 'true' : 'false').padStart(8)}`
        );
    });
}

main();
